### PR TITLE
implemented spring security in project

### PR DESCRIPTION
**Pull request by ynik404**

### CODE REVIEW

File: src/main/java/mate/academy/model/Role.java [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+import jakarta.persistence.GeneratedValue;
+import jakarta.persistence.GenerationType;
+import jakarta.persistence.Id;
+import jakarta.persistence.Table;
+import org.springframework.security.core.GrantedAuthority;
+
+@Entity
+@Table(name = "roles")
+public class Role implements GrantedAuthority {
+    private static final String ROLE_PREFIX = "ROLE_";
```
Comment (MentorMateAcademy) [inline src/main/java/mate/academy/model/Role.java:16]: why do you need this prefix?

**Review by MentorMateAcademy — CHANGES_REQUESTED** [review]

File: src/main/java/mate/academy/controller/BookController.java [hunk 2]
```diff
@@ -25,36 +27,48 @@
 @Tag(name = "Book management", description = "Endpoints for managing books")
 @RequiredArgsConstructor
 @RestController
-@RequestMapping("/books")
+@RequestMapping("/api/books")
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/controller/BookController.java:30]:
```suggestion
@RequestMapping("/books")
```

Add `server.servlet.context-path=/api` to application.properties

File: src/main/java/mate/academy/dto/user/UserRegistrationRequestDto.java [hunk 3]
```diff
@@ -13,7 +13,7 @@
 public class UserRegistrationRequestDto {
     @NotBlank
     @Email
-    private String email;
+    private String username;
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/dto/user/UserRegistrationRequestDto.java:16]: Why do you need to change it?

File: src/main/java/mate/academy/dto/user/UserResponseDto.java [hunk 4]
```diff
@@ -5,7 +5,7 @@
 @Data
 public class UserResponseDto {
     private Long id;
-    private String email;
+    private String username;
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/dto/user/UserResponseDto.java:8]: same

File: src/main/java/mate/academy/model/Role.java [hunk 5]
```diff
@@ -0,0 +1,33 @@
...
+import jakarta.persistence.GeneratedValue;
+import jakarta.persistence.GenerationType;
+import jakarta.persistence.Id;
+import jakarta.persistence.Table;
+import org.springframework.security.core.GrantedAuthority;
+
+@Entity
+@Table(name = "roles")
+public class Role implements GrantedAuthority {
+
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/model/Role.java:16]:
```suggestion
```

File: src/main/java/mate/academy/controller/BookController.java [hunk 6]
```diff
@@ -25,36 +27,48 @@
...
     @ResponseStatus(HttpStatus.CREATED)
     @Operation(summary = "Create a new book", description = "Create a new book")
     public BookDto createBook(@RequestBody @Valid CreateBookRequestDto requestDto) {
         return bookService.createBook(requestDto);
     }
 
+    @PreAuthorize("hasRole('ROLE_ADMIN')")
+    @PutMapping("/{id}")
+    @Operation(summary = "Change book's description", description = "Change book's description")
+    public void changeBookDescription(@PathVariable Long id, String description) {
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/controller/BookController.java:59]: It's better to create update method, to update not only description

File: src/main/java/mate/academy/security/CustomUserDetailsService.java [hunk 7]
```diff
@@ -0,0 +1,23 @@
...
+
+@Service
+@RequiredArgsConstructor
+public class CustomUserDetailsService implements UserDetailsService {
+    private final UserRepository userRepository;
+
+    @Override
+    public UserDetails loadUserByUsername(String email)
+            throws UsernameNotFoundException {
+        return userRepository.findByUsername(email)
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/security/CustomUserDetailsService.java:18]:
```suggestion
        return userRepository.findByEmail(email)
```

**Review by Elena-Bruyako — CHANGES_REQUESTED** [review]
see comments

File: src/main/java/mate/academy/controller/BookController.java [hunk 8]
```diff
@@ -29,32 +31,44 @@
...
     @ResponseStatus(HttpStatus.CREATED)
     @Operation(summary = "Create a new book", description = "Create a new book")
     public BookDto createBook(@RequestBody @Valid CreateBookRequestDto requestDto) {
         return bookService.createBook(requestDto);
     }
 
+    @PreAuthorize("hasRole('ROLE_ADMIN')")
+    @PutMapping("/{id}")
+    @Operation(summary = "Change book's description", description = "Change book's description")
+    public void updateBookById(@PathVariable Long id, CreateBookRequestDto requestDto) {
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/controller/BookController.java:59]:
```suggestion
    public BookDto updateBook(@PathVariable Long id, CreateBookRequestDto requestDto) {
```

**Review by Elena-Bruyako — CHANGES_REQUESTED** [review]
See comment

File: src/main/java/mate/academy/security/CustomUserDetailsService.java [hunk 9]
```diff
@@ -0,0 +1,23 @@
...
+@RequiredArgsConstructor
+public class CustomUserDetailsService implements UserDetailsService {
+    private final UserRepository userRepository;
+
+    @Override
+    public UserDetails loadUserByUsername(String email)
+            throws UsernameNotFoundException {
+        return userRepository.findByEmail(email)
+                .orElseThrow(()
+                        -> new UsernameNotFoundException("Can't find user by email: "
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/security/CustomUserDetailsService.java:20]:
```suggestion
                        -> new EntityNotFoundException("Can't find user by email: "
```

File: src/main/java/mate/academy/service/impl/BookServiceImpl.java [hunk 10]
```diff
@@ -56,4 +56,15 @@ public List<BookDto> search(BookSearchParameters params, Pageable pageable) {
                 .map(bookMapper::toDto)
                 .toList();
     }
+
+    @Override
+    public BookDto updateBookById(Long id, CreateBookRequestDto requestDto) {
+        Book book = bookRepository.findById(id)
+                .orElseThrow(() -> new EntityNotFoundException("Can't find book by id: " + id));
+        Book updatedBook = bookMapper.toModel(requestDto);
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/service/impl/BookServiceImpl.java:64]:
```suggestion
        bookMapper.updateBookFromDto(bookDto, book);
```

File: src/main/java/mate/academy/service/impl/BookServiceImpl.java [hunk 11]
```diff
@@ -56,4 +56,15 @@ public List<BookDto> search(BookSearchParameters params, Pageable pageable) {
...
+        book.setTitle(updatedBook.getTitle());
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/service/impl/BookServiceImpl.java:65]: use mapstruct for this logic

**Review by Elena-Bruyako — CHANGES_REQUESTED** [review]
Good job! Let's improve your solution

File: src/main/java/mate/academy/service/impl/UserServiceImpl.java [hunk 12]
```diff
@@ -24,7 +31,17 @@ public UserResponseDto register(UserRegistrationRequestDto requestDto)
                     .format("User with this email: %s already exists",
                             requestDto.getEmail()));
         }
-        User savedUser = userRepository.save(userMapper.toModel(requestDto));
+        User user = userMapper.toModel(requestDto);
+        user.setPassword(passwordEncoder.encode(requestDto.getPassword()));
+        user.setRoles(Set.of(
+                findRoleByName(Role.RoleName.USER)));
+        User savedUser = userRepository.save(user);
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/service/impl/UserServiceImpl.java:38]:
```suggestion
        userRepository.save(user);
```

File: src/main/java/mate/academy/service/impl/UserServiceImpl.java [hunk 13]
```diff
@@ -24,7 +31,17 @@ public UserResponseDto register(UserRegistrationRequestDto requestDto)
...
+
         return userMapper.toUserResponse(savedUser);
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/service/impl/UserServiceImpl.java:40]:
```suggestion
        return userMapper.toUserResponse(user);
```

File: src/main/java/mate/academy/model/Role.java [hunk 14]
```diff
@@ -0,0 +1,32 @@
...
+import jakarta.persistence.Entity;
+import jakarta.persistence.EnumType;
+import jakarta.persistence.Enumerated;
+import jakarta.persistence.GeneratedValue;
+import jakarta.persistence.GenerationType;
+import jakarta.persistence.Id;
+import jakarta.persistence.Table;
+import org.springframework.security.core.GrantedAuthority;
+
+@Entity
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/model/Role.java:13]: add get and set

**Review by Elena-Bruyako — CHANGES_REQUESTED** [review]
see pr

**Review by Elena-Bruyako — APPROVED** [review]
well done

**Pull request merged**